grafana/grafana
# Loki: Fix unique log row id generation

( fixes https://github.com/grafana/grafana/issues/53715 )

when we run a Loki query in the grafana-server, afterwards we build the dataframes that we send back to the browser. there is a field named `id` that is in the dataframes. the `id` should contain the `query.refId` attribute too, but it did not contain it, because of this bug.

(explanation: when we build the `Frame` objects in `go`, we do not have to fill out their `RefId` attribute, it will get filled out automatically by the go-server. but, in turn, we cannot rely on it being there. so instead of `frame.RefId`, we use `query.RefId`)

how to test:
- `make devenv sources=loki`
- go to explore in grafana, choose the loki datasource
- you get an empty query section with name `A`. rename it to `first_query`
- open the browser-devtools, make sure you can see the ajax requests
- run any loki query, for example `{place="moon"}`
- you will get two `query` ajax requests, one is for the logs, another for the log-volume-histogram , we want to see the logs-query, not the logs-histogram-query (check what was sent to the grafana-server,  you want to check the `queries[0].expr` field. it should contain our query, and not some kind of `sum by (level)...`)
- check the returned ajax-data, you want to navigate to `results.first_query.frames[0].data.values`, check the last item of this array. it should be an array of strings.these are the `id`s for the log-rows. all of them should end with `_first_query`
